docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,489 @@
+% kubectl get pods
+NAME          READY     STATUS    RESTARTS   AGE
+neato-5thn0   1/1       Running   0          1d
+% kubectl debug -it -m debian -p neato-5thn0 -- bash
+root@debug-image-neato-5thn0:/# cat /etc/resolv.conf
+search default.svc.cluster.local svc.cluster.local cluster.local
+nameserver 10.155.240.10
+options ndots:5
+root@debug-image-neato-5thn0:/# dig @10.155.240.10 neato.svc.cluster.local.
+
+; <<>> DiG 9.9.5-9+deb8u6-Debian <<>> @10.155.240.10 neato.svc.cluster.local.
+; (1 server found)
+;; global options: +cmd
+;; connection timed out; no servers could be reached
+```
+
+This leads Samantha to discover that the cluster's DNS service isn't responding.
+
+### Automation
+
+Abe is a security engineer tasked with running security audits across all of his
+company's running containers. Even though his company has no standard base
+image, he's able to audit all containers using:
+
+```
+% for pod in $(kubectl get -o name pod); do
+    kubectl debug -m gcr.io/neato/security-audit -p $pod /security-audit.sh
+  done
+```
+
+## Alternatives Considered
+
+### Mutable Pod Spec
+
+Rather than adding an operation to have Kubernetes attach a pod we could instead
+make the pod spec mutable so the client can generate an update adding a
+container. `SyncPod()` has no issues adding the container to the pod at that
+point, but an immutable pod spec has been a basic assumption in Kubernetes thus
+far and changing it carries risk. It's preferable to keep the pod spec immutable
+as a best practice.
+
+### Ephemeral container
+
+An earlier version of this proposal suggested running an ephemeral container in
+the pod namespaces. The container would not be added to the pod spec and would
+exist only as long as the process it ran. This has the advantage of behaving
+similarly to the current kubectl exec, but it is opaque and likely violates
+design assumptions. We could add constructs to track and report on both
+traditional exec process and exec containers, but this would probably be more
+work than adding to the pod spec. Both are generally useful, and neither
+precludes the other in the future, so we chose mutating the pod spec for
+expedience.
+
+### Attaching Container Type Volume
+
+Combining container volumes ([#831](https://issues.k8s.io/831)) with the ability
+to add volumes to the pod spec would get us most of the way there. One could
+mount a volume of debug utilities at debug time. Docker does not allow adding a
+volume to a running container, however, so this would require a container
+restart. A restart doesn't meet our requirements for troubleshooting.
+
+Rather than attaching the container at debug time, kubernetes could always
+attach a volume at a random path at run time, just in case it's needed. Though
+this simplifies the solution by working within the existing constraints of
+`kubectl exec`, it has a sufficient list of minor limitations (detailed in
+[#10834](https://issues.k8s.io/10834)) to result in a poor user experience.
```

Comment:
I'm posting an alternate workflow for debug volumes here just to facilitate a discussion. It is possible that you have already considered this idea and I have failed to identify the disadvantages with such an approach in your comments on #10834. Kindly bear with me.

1. It is possible to export a docker image into a volume using `docker export`
2. Such a volume can be injected into a container at runtime using mount propagation
3. An exec session in a container can then use the tools within the new debug volume. To work around PATH issues, the exec session can `chroot` assuming `chroot` is available in the debug volume. To have access to the container's root filesystem, the container's root filesystem can be mounted into the volume at a special path.
4. Once the special exec session ends, the debug volume can also be cleaned up.